### ISSUE TITLE
Run `go mod tidy` in Go projects

### ISSUE THREAD

**Issue by TheRedSpy15** [dependabot/dependabot-core#2229]
This would be awesome to have as a Go developer

Rather than just updating modules, running `go mod tidy` would add and remove them too

**Comment by siavashs** [comment]
We use `-mod=readonly` in our workflows and get this error on the PRs opened by the bot:
```
go: updates to go.sum needed, disabled by -mod=readonly
```

**Comment by aclevername** [comment]
> We use `-mod=readonly` in our workflows and get this error on the PRs opened by the bot:
> 
> ```
> go: updates to go.sum needed, disabled by -mod=readonly
> ```

same here! The only solution for us so far is to manually run a `go mod tidy` on the same branch as the PR, taking away from the lovely automation that we want 😢

**Comment by jurre** [comment]
We're starting to roll out this feature on the GH native version of Dependabot, if anyone would like to preview this, please let me know which repo or account you'd like it enabled on.

**Comment by jurre** [comment]
https://github.blog/changelog/2020-10-19-dependabot-go-mod-tidy-and-vendor-support/ 🎉

**Comment by dfarrell07** [comment]
Is this expected to be enabled for all repositories now? As best I can tell, we're not seeing it yet.

For example, [this PR from Dependabot](https://github.com/submariner-io/submariner-operator/pull/776) didn't include [these `go mod tidy` updates](https://github.com/submariner-io/submariner-operator/pull/837). The same for [this version bump](https://github.com/submariner-io/lighthouse/pull/352) and [these `go mod tidy` changes](https://github.com/submariner-io/lighthouse/pull/372).

Thanks for this great tooling, we really find it useful and appreciate it!

**Comment by dfarrell07** [comment]
> Is this expected to be enabled for all repositories now? As best I can tell, we're not seeing it yet.
> 
> For example, [this PR from Dependabot](https://github.com/submariner-io/submariner-operator/pull/776) didn't include [these `go mod tidy` updates](https://github.com/submariner-io/submariner-operator/pull/837) (or [this Dependabot PR from last week](https://github.com/submariner-io/submariner-operator/pull/7930)). The same for [this version bump](https://github.com/submariner-io/lighthouse/pull/352) and [these `go mod tidy` changes](https://github.com/submariner-io/lighthouse/pull/372).
> 
> Thanks for this great tooling, we really find it useful and appreciate it!

It turns out we needed an admin to login to the Dependabot dashboard and have the bot generate PRs like https://github.com/submariner-io/submariner/pull/1033 to update to the new GH-driven version. Thanks!

**Comment by jurre** [comment]
> > Is this expected to be enabled for all repositories now? As best I can tell, we're not seeing it yet.
> > For example, [this PR from Dependabot](https://github.com/submariner-io/submariner-operator/pull/776) didn't include [these `go mod tidy` updates](https://github.com/submariner-io/submariner-operator/pull/837) (or [this Dependabot PR from last week](https://github.com/submariner-io/submariner-operator/pull/7930)). The same for [this version bump](https://github.com/submariner-io/lighthouse/pull/352) and [these `go mod tidy` changes](https://github.com/submariner-io/lighthouse/pull/372).
> > Thanks for this great tooling, we really find it useful and appreciate it!
> 
> It turns out we needed an admin to login to the Dependabot dashboard and have the bot generate PRs like [submariner-io/submariner#1033](https://github.com/submariner-io/submariner/pull/1033) to update to the new GH-driven version. Thanks!

Happy you got it to work @dfarrell07, I'm slightly surprised it didn't work for dependabot-preview, but if you're able to use the GH native version that's much better 🎉

**Comment by juanibiapina** [comment]
This doesn't work for dependabot-preview in our case as well. `go mod tidy` is not run.

But this is not really the root of the problem. When running `go mod tidy` it will by side effect add a line that it's actually missing when updating:

Dependabot adds this line:
```
github.com/google/uuid v1.1.4/go.mod h1:TIyPZe4MgqvfeYDBFedMoGGpEw/LqOeaOT+nhxU+yHo=
```

But misses this line:
```
github.com/google/uuid v1.1.4 h1:0ecGp3skIrHWPNGPJDaBIghfA6Sp7Ruo2Io8eLKzWm0=
```

Then running `go build` (or any other `go` command) locally fixes it. Running `go mod tidy` also fixes it, but not because it removed the older versions.

In summary, looks like dependabot preview is completely broken currently, unless I understand this wrong.

**Comment by sylr** [comment]
I can confirm @juanibiapina statements.

PRs opened by dependabot do not pass my CI pipeline due to missing lines which would be generated by `go mod tidy`.

See: https://github.com/sylr/prometheus-azure-exporter/pull/190

**Comment by jurre** [comment]
> I can confirm @juanibiapina statements.
> 
> PRs opened by dependabot do not pass my CI pipeline due to missing lines which would be generated by `go mod tidy`.
> 
> See: [sylr/prometheus-azure-exporter#190](https://github.com/sylr/prometheus-azure-exporter/pull/190)

I'd recommend upgrading to the GH native version of dependabot. If you log into dependabot.com there should be a button that handles the migration for you. Go mod tidy should run as expected there 👍

**Comment by sylr** [comment]
Thank you @jurre! Everything is good now 👍

**Comment by sylr** [comment]
@jurre it seems that gotip (go built on the master branch) expect more cleanup that previous versions.

See: https://github.com/sylr/prometheus-azure-exporter/pull/192/checks?check_run_id=1836270903

**Comment by jurre** [comment]
> @jurre it seems that gotip (go built on the master branch) does more cleanup than the previous versions.
> 
> See: https://github.com/sylr/prometheus-azure-exporter/pull/192/checks?check_run_id=1836270903

Hmm I see, we won't be able to update until it's released as stable. What would you expect Dependabot to do in this case?

**Comment by sylr** [comment]
I don't know. I've tested and `go mod tidy` with gotip and it produces valid go.mod and go.sum for at least 1.13, 1.14 and 1.15.

**Comment by jurre** [comment]
> I don't know. I've tested `go mod tidy` with gotip and it produces valid go.mod and go.sum for at least 1.13, 1.14 and 1.15.

That's good info, thanks! Going to hold of until there's a stable release and then bump our go version 👍

**Comment by sylr** [comment]
Would it be possible to select which go version does the `go mod tidy` in `.github/dependabot.yml` ?

**Comment by jurre** [comment]
> Would it be possible to select which go version does the `go mod tidy` in `.github/dependabot.yml` ?

Not currently and it would be a ton of work to add it. I can definitely see why this would be nice, so maybe in the future but we probably won't have time for it in the foreseeable future. I guess when we do implement this we'd look at the version specified in the `go.mod` vs a separate config.

So far though the go versions have been backwards compatible so there's not been a strong need to be able to specify versions.

**Comment by aanm** [comment]
>So far though the go versions have been backwards compatible so there's not been a strong need to be able to specify versions.

@jurre can we re-open this issue? It seems that go-mod 1.16 is more aggressive cleaning up entries from `go.sum` compared to older versions which then makes the PRs where the dependencies are being updated to fail in the go-mod checkers.

**Comment by owais** [comment]
Dependabot is currently unusable for out Go projects  because of this.

https://github.com/signalfx/signalfx-go-tracing/pull/71
https://github.com/signalfx/signalfx-go-tracing/pull/75

These PR would be usable if dependabot ran `go mod tidy` before submitting the PRs.

**Comment by owais** [comment]
Dependabot should have a hook system in place. It should allow users to specify custom programs to call instead of the default ones for the language. So instead of `npm install <dep>` for Node, it would call `./scripts/update_dep.sh <dep>` if specified by the user. This would allow users to fix such cases themselves as needed.

**Comment by jurre** [comment]
> These PR would be usable if dependabot ran go mod tidy before submitting the PRs.

We do run `go mod tidy`, but it appears that your project is still on `go 1.12` which handles `go mod tidy` slightly differently. Dependabot runs go 1.16, and have some code in place to make sure we don't swap out the go version in the `go.mod` file. I agree this is unfortunate and at some point we would like to support multiple versions of the go toolchain at some point, but we're not there yet. 

> Dependabot should have a hook system in place. It should allow users to specify custom programs to call instead of the default ones for the language.

To be honest we're not likely to add this anytime soon, it would be very hard for us to support at scale and it's ripe for abuse. You can run a GitHub action after Dependabot runs to perform some custom steps though. You'll need to use [pull_request_target](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request_target) as the  trigger. Similar suggestions are discussed more [here](https://github.com/dependabot/dependabot-core/issues/1758) and [here](https://github.com/dependabot/dependabot-core/issues/2004), but it seems offtopic for this thread.

Some options I can offer to resolve this would be updating the go version used in the project, or setting up an action/ci job to run after Dependabot and commit the results to the branch.

**Comment by owais** [comment]
> We do run go mod tidy, but it appears that your project is still on go 1.12 which handles go mod tidy slightly differently. Dependabot runs go 1.16, and have some code in place to make sure we don't swap out the go version in the go.mod file. I agree this is unfortunate and at some point we would like to support multiple versions of the go toolchain at some point, but we're not there yet.

I'm pretty sure we run 1.16. My colleague pointed out that the issue is that this is a monorepo with a bunch of modules depending on the module at root of the repo. As a result, we need to run `go mod tidy` in every module even though the dep changed only in the root module. Not sure if running `go mod tidy` in all modules irrespective of whether a dep changed for them or not is safe to do or not. The bot could look for `replace ` directives refering to the root module and only run it then but not sure if it makes sense to you to account for something like `replace`. 

I think it might be a good solution to run `go mod tidy` in every Go module found in a repo irrespective of whether all of them were updated or not. I just don't know if there can be any downsides or annoyances to this. I can't think of any.

**Comment by crazy-max** [comment]
@jurre

> To be honest we're not likely to add this anytime soon, it would be very hard for us to support at scale and it's ripe for abuse. You can run a GitHub action after Dependabot runs to perform some custom steps though. You'll need to use [pull_request_target](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request_target) as the trigger. Similar suggestions are discussed more [here](https://github.com/dependabot/dependabot-core/issues/1758) and [here](https://github.com/dependabot/dependabot-core/issues/2004), but it seems offtopic for this thread.

I've created [a GitHub Actions workflow to fix dependabot issues](https://github.com/crazy-max/artifactory-cleanup/blob/master/.github/workflows/vendor.yml) with go.sum but looks overkill to me:

![image](https://user-images.githubusercontent.com/1951866/115367726-bf4ca580-a1c6-11eb-8129-223042306710.png)

I also need to define a PAT to be able to re-trigger workflows after push through this workflow. So not a final solution for me. Would be nice to be able to define go version in dependabot yaml config.

**Comment by owais** [comment]
> To be honest we're not likely to add this anytime soon, it would be very hard for us to support at scale and it's ripe for abuse. 

Yeah, I figured that.

> You can run a GitHub action after Dependabot runs to perform some custom steps though. You'll need to use pull_request_target as the trigger. Similar suggestions are discussed more here and here, but it seems offtopic for this thread.

This might be a viable solution. My concern with it is that Github Actions will need write access to the repo or the dependabot fork (if it doesn't already) which opens a whole another can of worms (codecov breach would be able to write to repos for example). 

@pellared something we can consider perhaps? ^ Let's discuss offline.

**Comment by jurre** [comment]
> I'm pretty sure we run 1.16. 

https://github.com/signalfx/signalfx-go-tracing/blob/dc035b573bfb2934c0c010b543fe0d7cab76c2c3/go.mod#L3 seems to indicate you're on v 1.12? Might not matter though

> As a result, we need to run go mod tidy in every module even though the dep changed only in the root module

Ah, yeah Dependabot currently only runs `go mod tidy` in the root module. I think running in every submodule _might_ be possible, but I'm worried it might be a lot of work, and I'm not sure how common this scenario is.

I did run Dependabot on a clone of the repo and made sure we do run `go mod tidy` for this project (we do) and that running `go mod tidy` manually after the update doesn't give us any other changes and that all looks good. 

Do you have some sort of script you run to run `go mod tidy` on every required folder that I can verify against?

**Comment by jurre** [comment]
> Would be nice to be able to define go version in dependabot yaml config.

@crazy-max I agree, it would be even nicer if Dependabot just figured out the right go version from the `go.mod` file and used that. 

I think eventually we'll want to support this, but I can't offer a timeline right now. Most of the golang ecosystem seems to stay fairly up to date with go versions, and backwards compatibility is usually considered thoughtfully in version upgrades, but still agree this would be nice.

**Comment by crazy-max** [comment]
@jurre 

> I agree, it would be even nicer if Dependabot just figured out the right go version from the `go.mod` file and used that.

[`go` directive](https://golang.org/ref/mod#go-mod-file-go) in `go.mod` should not be used to determine the current Go version of a project:

> A go directive indicates that a module was written assuming the semantics of a given version of Go.

For example a project compiled with Go 1.16 may well have a `go 1.13` directive in its `go.mod` to indicated backwards compatibility of its code. That's why I suggest instead smth in your dependabot yml config.
